pysph/cpy/jit.py

```diff
@@ -0,0 +1,523 @@
+    unsigned_a = unsigned_b = False
+    if a.startswith('u'):
+        unsigned_a = True
+        a = a[1:]
+    if b.startswith('u'):
+        unsigned_b = True
+        b = b[1:]
+    idx_a = preference_order.index(a)
+    idx_b = preference_order.index(b)
+    return_type = preference_order[idx_a] if idx_a > idx_b else \
+        preference_order[idx_b]
+    if unsigned_a and unsigned_b:
+        return_type = 'u%s' % return_type
+    return return_type
+
+
+class AnnotationHelper(ast.NodeVisitor):
+    def __init__(self, func, arg_types):
+        self.func = func
+        self.arg_types = arg_types
+        self.var_types = arg_types.copy()
+        self.children = {}
+
+    def get_type(self, type_str):
+        kind, address_space, ctype, shape = get_declare_info(type_str)
+        if 'unsigned' in ctype:
+            ctype = ctype.replace('unsigned ', 'u')
+        if kind == 'matrix':
+            ctype = '%sp' % ctype
+        return ctype
+
+    def annotate(self):
+        src = dedent('\n'.join(inspect.getsourcelines(self.func)[0]))
+        self._src = src.splitlines()
+        code = ast.parse(src)
+        self.visit(code)
+        self.func = types.annotate(self.func, **self.arg_types)
+
+    def error(self, message, node):
+        msg = '\nError in code in line %d:\n' % node.lineno
+        if self._src:  # pragma: no branch
+            if node.lineno > 1:  # pragma no branch
+                msg += self._src[node.lineno - 2] + '\n'
+            msg += self._src[node.lineno - 1] + '\n'
+            msg += ' ' * node.col_offset + '^' + '\n\n'
+        msg += message
+        raise NotImplementedError(msg)
+
+    def warn(self, message, node):
+        msg = '\nIn code in line %d:\n' % node.lineno
+        if self._src:  # pragma: no branch
+            if node.lineno > 1:  # pragma no branch
+                msg += self._src[node.lineno - 2] + '\n'
+            msg += self._src[node.lineno - 1] + '\n'
+            msg += ' ' * node.col_offset + '^' + '\n\n'
+        msg += message
+        warnings.warn(msg)
+
+    def visit_Call(self, node):
+        # FIXME: External functions have to be at the module level
+        # for this to work. Pass list of external functions to
+        # make this work
+        mod = importlib.import_module(self.func.__module__)
+        f = getattr(mod, node.func.id, None)
+        if not hasattr(f, 'is_jit'):
+            return None
+        if node.func.id in self.children:
+            return self.children[node.func.id].arg_types.get(
+                'return_', None)
+        if isinstance(node.func, ast.Name) and \
+                node.func.id not in BUILTINS:
+            if f is None or isinstance(f, Extern):
+                return None
+            else:
+                arg_types = []
+                for arg in node.args:
+                    arg_type = self.visit(arg)
+                    if not arg_type:
+                        msg = "Function called is not marked by the jit "\
```

Comment:
Use a triple quoted string with `textwrap.dedent` put brackets to avoid the `\` at the end of the line?